Bulk delete drafts

juliaBichler01 opened the pull request:
Resolves zulip/zulip#19360

A added the possibility to bulk-delete drafts (instead of having to delete each one individually):
![bulk-delete_all](https://user-images.githubusercontent.com/74348920/140652397-0bda523c-a906-4e2e-bfde-ea04de3b34c2.gif)

It is also possible to mark the drafts individually or to unselect all drafts:
![bulk-delete_markManually](https://user-images.githubusercontent.com/74348920/140652465-f6a0f256-e5ce-4fad-9322-f935c1d53b03.gif)

And you can unselect specifc drafts so you can delete all except those drafts:
![bulk-delete_notAll](https://user-images.githubusercontent.com/74348920/140652552-6a6c2aa2-e782-41ab-9a9c-98229edb4791.gif)

Testing:
Manually, might need some formal tests in frontend_tests/node_tests or frontend_tests/puppeteer_tests

juliaBichler01 commented:
I think this is looking pretty good now, so a review would be great :)
(To avoid any confusion: I am not completely sure if it's better to ask for a review here or on the community server, so i did both)

timabbott commented:
@juliaBichler01 thanks for working on this!  The UI looks reasonable.  Can you squash your commits?  That's made this branch much more readable.  
Check out our [GitHub guide](https://zulip.readthedocs.io/en/latest/git-guide.html) and [commit guidelines](https://zulip.readthedocs.io/en/latest/contributing/version-control.html) for details and advice on doing this.

I think this only does one of the two requests in the issue: checkboxes to deletes specific sets of drafts.  Do you want to do a separate pull request to add a little sidebar menu option to delete all drafts?  That should be very doable now that https://github.com/zulip/zulip/pull/18819 is merged.

juliaBichler01 commented:
Okay, i squashed my commits.
I think making a separate pull request for the other part would be a good idea, i would be happy to do that!
I still have one question about this pull request though: should i also add some formal tests for this feature?

timabbott commented:
@juliaBichler01 re: tests -- I'd look at the node test suite for drafts (and read the docs on writing node tests) and see if you can figure out a good way to do them.

juliaBichler01 commented:
@timabbott it seems like the frontend node tests should primarily test changes in the data, did i understand that correctly?
Because then i am not sure if i even should add tests, because i did not implement any new functions that do that, i only added changes to the appearance and reused the existing remove function to delete the drafts?

timabbott commented:
Yes, that's correct -- we generally aim to have the frontend node tests focus on testing the data aspects of things.  If one needs a UI test, one can do that with puppeteer, but we try to avoid adding those unless there's a very good reason to, since they have a substantial maintenance burden.

juliaBichler01 commented:
@timabbott okay, then i think it is not really beneficial to add tests, as the only changes to the data that happen are done by using the draft_model.deleteDraft() function, and this is already tested.
Is there something else that has to be done before this pull request can be merged?

COM8 commented:
@timabbott is there anything else we have to wait before merging this PR, since this one is especially handy if you use drafts like me for notes or as a quick and dirty To-Do list.

alya commented:
@juliaBichler01 I think it would be worth posting some screenshots in #design for feedback to fine-tune the UI. I feel like the horizontal alignment ends up looking a bit odd (both when the checkboxes are selected and when they are not), but I'm not sure how best to fix it.

A couple of other comments:
* The borders for unselected checkboxes (seen on hover) look too dark to me in light mode. Perhaps we can make them match the settings checkboxes?
* The unselected checkboxes don't look right in dark mode:
![Screen Shot 2022-01-06 at 8 54 21 AM](https://user-images.githubusercontent.com/2090066/148420189-c51911fd-cba5-40a0-a378-a0186cc5bdcd.png)

Thanks for working on this feature -- it'll be really handy!

alya commented:
Thanks for updating!

Hm, dark mode checkboxes still don't look right to me. Are you using the same ones as we have in the settings menus?

I guess there's an additional complication that they have to be visible with the lighter background we use for PMs. Currently, they are almost impossible to see:
![Screen Shot 2022-02-04 at 10 07 29 AM](https://user-images.githubusercontent.com/2090066/152581261-3205b01b-d1b1-4348-b430-255d6ab9a15d.png)

alya commented:
The responsive behavior looks fine for a while, but eventually breaks down when you get to a really narrow screen:
![Screen Shot 2022-02-04 at 10 17 49 AM](https://user-images.githubusercontent.com/2090066/152582022-656661e3-4f80-4891-a53d-3c48bff5e012.png)

alya commented:
Could you please post screenshots with "(Un)Select all drafts" selected and unselected, in dark and light mode? I think it would be worth getting feedback on how the pill looks (and whether we want it look like a pill at all), as well as styling for the top checkbox.

juliaBichler01 commented:
@alya thanks for the review!
I am using the same checkboxes as in the settings, but i reduced the opacity and only put it on 100% on hover. I had the opacity on 20% until now, i guess that's a little bit too less. I changed it to 40% now, though.
I also fixed the issue for really narrow windows.
As requested, i posted some screenshots of all of this on the community server, you can see if you like it there :)

juliaBichler01 commented:
@alya I think this is ready for a review now!

alya commented:
Cool, thanks!

Are the checkboxes on individual drafts the same color as our settings checkboxes? I'm finding them really hard to see; I think making them lighter would look better.

<img width="78" alt="Screen Shot 2022-03-09 at 10 58 53 AM" src="https://user-images.githubusercontent.com/2090066/157513069-dd7e1339-f58c-4bd3-b4a5-7a9e18b1f331.png">

alya commented:
Otherwise looks good to me from a UI perspective!

alya commented:
@m-e-l-u-h-a-n would you be up for reviewing this one?

juliaBichler01 commented:
> Are the checkboxes on individual drafts the same color as our settings checkboxes in dark mode? I'm finding them really hard to see; I think making them lighter would look better.

@alya theses checkboxes are so light because i set the opacity to 0.4 and to 1 on hover so that there is a visual difference when user hover over them.
But i guess 0.4 was a little bit too light, so i changed it to 0.7 to match the values of the delete and restore buttons. I hope that this looks better!

juliaBichler01 commented:
@m-e-l-u-h-a-n That would be great if you could review this, thanks in advance!

m-e-l-u-h-a-n commented on static/templates/draft.hbs:
```diff
@@ -35,6 +35,10 @@
                             <i class="fa fa-pencil fa-lg restore-draft tippy-zulip-tooltip" aria-hidden="true" data-tippy-content="{{t 'Restore draft' }} (Enter)"></i>
                             <i class="fa fa-trash-o fa-lg delete-draft tippy-zulip-tooltip" aria-hidden="true" data-tippy-content="{{t 'Delete draft' }} (Backspace)"></i>
+                            <label class="checkbox">
+                                <input type="checkbox" id="mark-draft" class="mark-draft" aria-hidden="true"/>
```
`id` attribute here would be same for every draft which it shouldn't be ideally. I guess this is not required as `#mark-draft` could easily be replaced with its class selector `.mark-draft`.

m-e-l-u-h-a-n commented on static/styles/drafts.css:
```diff
@@ -46,6 +46,53 @@
             }
         }
+
+        .select-drafts-button {
+            position: absolute;
+            right: 35px;
+            top: 70px;
+            padding-right: 40px;
+            padding-top: 8px;
+
+            .fa-check-square {
+                position: absolute;
+                font-size: 18px;
+                top: 10px;
+                right: 7px;
+            }
+
+            .fa-square-o {
+                position: absolute;
+                font-size: 20px;
+                top: 9px;
+                right: 7px;
+            }
+
+            .button-text {
+                font-size: 14px;
+                padding-left: 8px;
+            }
+
+            .select-state-indicator {
+                padding-left: 5px;
+                padding-right: 8px;
+            }
+        }
+
```
I think we are setting a lot of values position related values for `.delete-drafts-button` and `.select-drafts-button` in a hard coded manner by using `position: absolute`.  Which is probably the reason for position of select all drafts button to be unstable.

A more stable way to fix this could be to have drafts-header and drafts in two separate containers (divs) so that scrollbar only appears for drafts list and not for drafts header. Similar to what we do in other stream/organization settings modals.

m-e-l-u-h-a-n reviewed (commented):
I had a look at this pr and have a added a few comments that are mostly over the fact that we are not using CSS in a stable and efficient manner here that's why we have to set a lot of styles/properties using `js` and this could cause a lot of future bugs in styling/positioning of elements here.
I think if we go about fixing this as following series of commits we would land on a more stable and clean fix.

- Refactor [static/templates/draft_table_body.hbs](https://github.com/zulip/zulip/blob/main/static/templates/draft_table_body.hbs) so that scroll bar only covers draft list area and not entire drafts modal.
- Left align a draft header content.
- Add bulk delete UI and refactor drafts header to have a stable positioning of elements (probably by using flex-box.)

m-e-l-u-h-a-n commented:
I think you should also add comments whenever you are doing some unconventional changes that you think might be unclear to someone going through that code later. That would improve the readability of the code and would also help in quick reviewing.
You can read more about it and commit disciplines at [https://zulip.readthedocs.io/en/latest/contributing/version-control.html](https://zulip.readthedocs.io/en/latest/contributing/version-control.html).

juliaBichler01 commented:
@m-e-l-u-h-a-n could you explain what you mean by "so that scroll bar only covers draft list area and not entire drafts modal"?
The scrollbar is never in the header?

m-e-l-u-h-a-n commented:
Ohh sorry I missed that, the scroll bar only covers the drafts list area. What I find peculiar is the right shift of delete draft button when most of the drafts are deleted and and scroll bar dissappears.
    <details>
        <summary>**GIF**</summary>
![Peek 2022-03-14 00-59](https://user-images.githubusercontent.com/63504956/158075875-d12d920d-2477-4910-af03-5543edf06296.gif)
    </details>   

so may be you could ignore first commit I suggested in [above comment](https://github.com/zulip/zulip/pull/19943#pullrequestreview-907684585), instead of that we can probably have a third commit that keeps the position of mark all drafts button stable.

juliaBichler01 commented:
@m-e-l-u-h-a-n regarding your last comment about keeping the position of the buttons stable:
It was decided in the community server that the right side of this button should align with the right side of the drafts ([discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/bulk.20delete.20drafts.20.2319943))
So to do that the button has to be shifted a bit sometimes to account for the space that the scrollbar takes up (or doesn't take up if there are just a few drafts).
Do you have an idea how i could solve this without moving the button?

m-e-l-u-h-a-n commented:
@juliaBichler01 regarding https://github.com/zulip/zulip/pull/19943#issuecomment-1080435259
As shifting of draft table elements is due to scroll bar appearence. I think it should be addressable by adding conditional `right-margin` for select all draft button in case of scroll bar, but that would simply complicate the style logic in this section. So I think that shifting of  that button is okay for now, and we can wait for Tim's review on the PR.

juliaBichler01 commented on static/js/drafts.js:
```diff
@@ -453,6 +487,46 @@ export function launch() {
 }
 
+function update_bulk_delete_ui() {
+    if ($(".drafts-list")[0].scrollHeight > $(".drafts-list")[0].clientHeight) {
+        $(".select-drafts-button").css("right", "35px");
+        $(".delete-drafts-button").css("right", "190px");
+    } else {
+        $(".select-drafts-button").css("right", "23px");
+        $(".delete-drafts-button").css("right", "178px");
+    }
```
i changed that, now this is a flex box as you reccomended

evykassirer replied:
Ah great this is after flexbox now! Is there a way you can do this with flexbox instead of setting the pixels manually?

juliaBichler01 commented:
@m-e-l-u-h-a-n okay, thanks for your feedback!

juliaBichler01 commented:
@timabbott could you review this Pr please?

evykassirer commented on static/templates/draft_table_body.hbs:
```diff
@@ -6,11 +6,13 @@
                     <span class="exit-sign">&times;</span>
                 </div>
-                <div class="removed-drafts">
+                <span class="removed-drafts">
```
The changes in this file seem like they were accidentally added to this commit! I'm guessing they're from a merge conflict where the wrong (old code-base) code was picked.

evykassirer commented on static/js/drafts.js:
```diff
@@ -428,6 +428,28 @@ function update_rendered_drafts(has_drafts_from_conversation, has_other_drafts)
     }
 }
 
+function updateButtonPlacement() {
```
It would also be great if this could be handled by flexbox or something similar, instead of manual pixel adjustments here.

juliaBichler01 replied:
this has to be changed after #21280 get merged anyway, so in my opinion changing this so it uses flexbox seems like a lot of unnecessary work, when it will have to be changed anyway
@alya what do you think?

evykassirer replied:
Can you explain more about how it will need to be changed? I'd ideally review this PR without it having the manual pixel adjustments, but maybe that means building it over that other PR?

evykassirer reviewed (commented):
I took a first look! I can look again more carefully once the flexbox commit is moved earlier, since I think that will change a handful of things in the other commits.

Also, some design thoughts:
* How do you feel about using "select" instead of "mark"?
* I'd also suggest keeping the trash icon there but have it greyed out when nothing is selected, otherwise I feel like it's unclear what the purpose of selecting drafts is

Thanks for working on this! Let me know if you have any questions about anything I said :)

alya commented:
> * How do you feel about using "select" instead of "mark"?
> * I'd also suggest keeping the trash icon there but have it greyed out when nothing is selected, otherwise I feel like it's unclear what the purpose of selecting drafts is

Both of those make sense to me!

juliaBichler01 commented on frontend_tests/node_tests/drafts.js:
```diff
@@ -445,11 +454,25 @@ test("format_drafts", ({override_rewire, mock_template}) => {
 
     $.create("#drafts_table .draft-row", {children: []});
+    let checkbox = $.create(".mark-draft");
+    checkbox.filter = () => [];
```
i don't really know why, but for some reason i need both

juliaBichler01 commented:
@alya @evykassirer i implemented most of the reviews and left comments on the two suggestions i didn't implement explaining why not

alya commented:
Thanks! Testing again, a few visual tweaks we should make:

1. The "Select all drafts" checkbox should be vertically aligned with the other checkboxes.

 ![Screen Shot 2023-01-23 at 11 45 45 AM](https://user-images.githubusercontent.com/2090066/214135034-4bf3297c-1def-4e50-a76b-c85b6160b590.png)
 
2. When I check/uncheck the "Select all drafts" box, the label on that button drifts up and down a bit. Can we make that not happen?
3. The "Select all drafts" looks a bit higher than the trash can button; they should be vertically aligned.
![Screen Shot 2023-01-23 at 10 59 36 AM](https://user-images.githubusercontent.com/2090066/214135413-8c55e9e7-bc46-406d-8288-b8a99b701df9.png)
4. Let's add a tooltip to the trash can button when it's disabled, to make it clear why. "No drafts selected" should be fine.
5. For the checkbox tooltip:
    a. Rather than saying "Select/Deselect draft", it should just say whichever one of those is appropriate, e.g., "Select draft" if it's currently unselected.
   b. Can we add a bit of a delay before showing this one? It's pretty obvious what the checkbox is for, so having it appear right away is unnecessarily distracting.

alya commented:
Also, the checkboxes look disabled to me in dark theme, especially when compared to the "Select all" checkbox. Maybe they should be colored like the "Select all" checkbox currently is?

![Screen Shot 2023-01-23 at 11 52 14 AM](https://user-images.githubusercontent.com/2090066/214136300-76f93400-3920-4597-b575-092efcc78e1e.png)

juliaBichler01 commented:
> 

I implemented all those comments

juliaBichler01 commented:
@evykassirer could you maybe review this again too? thanks in advance!

evykassirer commented on static/templates/draft_table_body.hbs:
```diff
@@ -6,10 +6,14 @@
+                        <br />
+                        {{#tr}}Drafts older than <strong>{draft_lifetime}</strong> days are automatically removed.{{/tr}}
+                        <br />
+                        {{t "Pro tip: You can use 'd' to open your drafts."}}
```
Can you remove this protip?

juliaBichler01 replied:
@evykassirer okay, i removed this line

alya commented:
Work on this PR is continued in #25610. Thanks @juliaBichler01 !
